### PR TITLE
Generate code coverage reports on PRs

### LINKED ISSUE #203: Implement code coverage reports on pull requests 🔍
To encourage contribution best practices, we should run coverage checks on critical logic in the library against our current test suite.

```[tasklist]
### Tasks
- [ ] Use `nyc` to generate test coverage reports
- [ ] Assess coverage reports on new code only (if possible)
- [ ] Implement a GitHub Action to present and block PRs if coverage on new logic is below 80%
- [ ] Rename test files to `*.test.js`
```

### PR DESCRIPTION
**Pull request by CyberCitizen01**
Closes #203

### CODE REVIEW

File: .github/workflows/nodejs.yml
Comment (JamieSlome) [inline .github/workflows/nodejs.yml:54]: @maoo - thoughts on these changes?

Comment (CyberCitizen01) [reply .github/workflows/nodejs.yml:54]: The context is that I've combined recording coverage and covered lines limit checking within the test step to avoid re-running all tests. The next step is to report the findings as a PR comment. Storing the exit code in the test step is essential because if the codebase doesn't meet the required criteria, it errors out. However, we need it to first report the findings as a PR comment, so I store the exit code and pass it to the last step. The workflow then exits and prevents the PR from merging.

You can see how this behaves and what data is presented at CyberCitizen01#1 (A test PR) and the [workflow run](https://github.com/CyberCitizen01/git-proxy/actions/runs/5707493190/job/15464227432).

Comment (JamieSlome) [reply .github/workflows/nodejs.yml:54]: @CyberCitizen01 - thanks for the detailed description 👍

**Comment by CyberCitizen01** [comment]
Hi
I've just noticed that the [latest workflow run](https://github.com/finos/git-proxy/actions/runs/5794787434/job/15705075109?pr=213) in this repository has failed. I've a hunch that it has something to do with permissions seeing the 403. Can I get a screenshot of the `Settings -> Action -> General -> Workflow permissions` page or maybe you can confirm if everything seems fine (maybe the action [`romeovs/lcov-reporter-action`]() needs to be added in the `Allow specified actions and reusable workflows` list?).
Thanks!

**Comment by JamieSlome** [comment]
@TheJuanAndOnly99 @maoo - I defer to you here. It appears the Action is not running as expected. Any ideas?

**Comment by TheJuanAndOnly99** [comment]
Hi @JamieSlome apologies for the delay. We're just getting back from some time off. We'll take a look as soon as we can and get back to you!

**Comment by JamieSlome** [comment]
@TheJuanAndOnly99 - no worries, thanks for the update, and hope you had a good break! 👍

**Comment by JamieSlome** [comment]
@CyberCitizen01 - in the meantime, could you take a look at the merge conflicts?

**Comment by TheJuanAndOnly99** [comment]
Hi @CyberCitizen01 @JamieSlome The setting seem fine. Action Permissions are set to allow all and workflow permissions are set to read and write and permissions. I may be wrong but it seems like you have configured it with a [file that does not exist](https://github.com/finos/git-proxy/pull/213/checks#step:8:6). Is that possible?

**Comment by CyberCitizen01** [comment]
@JamieSlome - I'll look into the merge conflicts right away.
@TheJuanAndOnly99 - Thanks for the update and your suggestion. I'll try to investigate this further and cross check if I've missed anything, although it seems to work for a dummy PR (CyberCitizen01#1) in my fork. So, it seems that GH possibly doesn't permit workflows in PRs to run with write permissions maybe?

**Comment by JamieSlome** [comment]
@CyberCitizen01 - thanks 👍

**Comment by JamieSlome** [comment]
@CyberCitizen01 - any updates here?

**Comment by CyberCitizen01** [comment]
Hi @JamieSlome apologies for the delay. I was caught up in my uni exams. I'll take a look as soon as possible and get back to you.

**Comment by JamieSlome** [comment]
@CyberCitizen01 - not a problem at all! I hope the exams went well 👍

**Comment by CyberCitizen01** [comment]
Hello there, sorry for being late.

I tried finding an appropriate GH action for "Assess coverage reports on new code only" https://github.com/finos/git-proxy/issues/203#issuecomment-1669194811 and at best I was able to discover this thread https://github.com/istanbuljs/nyc/issues/377. To implement this, I think I'll be required to create a script, maybe something similar to the one in scripts folder. Also I was not able to find a GH action that does this, and I'm still trying to find one (maybe I can get some help in finding one?). 

I would like to have some advice on how to move forward.

Thanks!

**Comment by coopernetes** [comment]
@CyberCitizen01 did some digging and I have a patch that should satisfy that requirement. Can you create a new file in your git repo (ie. "coverage.patch") using the contents below and use `git apply coverage.patch` to apply those changes to your branch and push? You should start seeing the `include` setting get populated based on the git diff output. Here's a working example: 
https://github.com/coopernetes/git-proxy/pull/2
https://github.com/coopernetes/git-proxy/actions/runs/6598811047/job/17927159890
https://github.com/coopernetes/git-proxy/compare/62dad6808ba799e925ac654699eff9fa467fdece..1fc80f340fc2d22843e437d3436b069e2262c018

```diff
diff --git a/.github/workflows/nodejs.yml b/.github/workflows/nodejs.yml
index 84e8ee0..6cb4630 100644
--- a/.github/workflows/nodejs.yml
+++ b/.github/workflows/nodejs.yml
@@ -24,6 +24,8 @@ jobs:
 
     steps:
     - uses: actions/checkout@8ade135a41bc03ea155e62e844d188df1ea18608 # v4
+      with:
+        fetch-depth: 0
 
     - name: Use Node.js ${{ matrix.node-version }}
       uses: actions/setup-node@v3
diff --git a/.nycrc.json b/.nycrc.json
deleted file mode 100644
index a61301d..0000000
--- a/.nycrc.json
+++ /dev/null
@@ -1,6 +0,0 @@
-{
-  "branches": 80,
-  "lines": 80,
-  "functions": 80,
-  "statements": 80
-}
diff --git a/README.md b/README.md
index 60a5e4c..6ac7c0d 100644
--- a/README.md
+++ b/README.md
@@ -290,3 +290,4 @@ This project is distributed under the Apache-2.0 license. See <a href="./LICENSE
 
 If you have a query or require support with this project, [raise an issue](https://github.com/finos/git-proxy/issues). Otherwise, reach out to [help@finos.org](mailto:help@finos.org).
 
+
diff --git a/nyc.config.js b/nyc.config.js
new file mode 100644
index 0000000..2971607
--- /dev/null
+++ b/nyc.config.js
@@ -0,0 +1,41 @@
+/* eslint-disable max-len */
+'use strict';
+
+const { execFileSync } = require('child_process');
+
+let opts = {
+  branches: 80,
+  lines: 80,
+  functions: 80,
+  statements: 80,
+};
+
+// Only generate coverage report for changed files in PR
+// see: https://github.com/actions/checkout/issues/438#issuecomment-1446882066
+//      https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
+if (process.env.GITHUB_BASE_REF !== undefined) {
+  console.log('Generating coverage report for changed files...');
+  try {
+    const baseRef = execFileSync('git', [
+      'rev-parse',
+      `origin/${process.env.GITHUB_BASE_REF}`,
+    ])
+      .toString()
+      .replace('\n', '');
+    const headRef = process.env.GITHUB_SHA;
+    const stdout = execFileSync('git', [
+      'diff',
+      '--name-only',
+      `${baseRef}..${headRef}`,
+    ]).toString();
+    opts = {
+      ...opts,
+      include: stdout.split('\n'),
+    };
+  } catch (error) {
+    console.log('Error: ', error);
+  }
+}
+
+console.log('nyc config: ', opts);
+module.exports = opts;
diff --git a/package.json b/package.json
index 1bcaa42..82e1a12 100644
--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     "server-test": "mocha --exit",
     "test": "mocha --exit",
     "test-coverage": "nyc npm run test",
-    "test-coverage-ci": "nyc --reporter=lcovonly --reporter=text-summary --check-coverage npm run test",
+    "test-coverage-ci": "nyc --reporter=lcovonly --reporter=text --check-coverage npm run test",
     "prepare": "node ./scripts/prepare.js",
     "lint": "eslint --fix . --ext .js,.jsx"
   },
```

File: .github/workflows/nodejs.yml
Comment (coopernetes) [inline .github/workflows/nodejs.yml:53]: @JamieSlome until we close #240, I would err on the side of removing this explicit failure in the workflow (maybe commenting out for now to be re-introduced after coverage is brought up to the thresholds?). 

For the time being, coverage can be reported in PRs for informational purposes.

Comment (JamieSlome) [reply .github/workflows/nodejs.yml:53]: @coopernetes, agreed 👍 No use blocking PRs and test creation should be incremental in any case.

**Comment by JamieSlome** [comment]
@coopernetes - really nice suggestion on the above. Once we get this applied to your PR, @CyberCitizen01, we can look to merge 🍰

**Comment by CyberCitizen01** [comment]
Hi @JamieSlome @coopernetes 

Thank you for the suggestion. I was thinking something similar. Rather I was thinking of using [`get-all-changed-files`](https://github.com/marketplace/actions/get-all-changed-files) action, but using `git diff --name-only` seems, neater! Thank you @coopernetes 

I'll apply the changes right away.

File: .github/workflows/nodejs.yml
Comment (CyberCitizen01) [reply .github/workflows/nodejs.yml:53]: Agreed.
I'll comment out the explicit failure part.

**Comment by JamieSlome** [comment]
@CyberCitizen01 - LGTM! 🍰 

Is it correct that no comment was made for this PR?

**Review by JamieSlome — COMMENTED** [review]
Nice job @CyberCitizen01 👍 Can you verify that the run of the job on this action had the expected outcome?

**Comment by CyberCitizen01** [comment]
It works as expected, I created a new PR https://github.com/CyberCitizen01/git-proxy/pull/3.

Some cases to better understand the intended behaviour (assuming the tasks of #240 are met):
- When a PR does involve modifications in those files for which, corresponding tests are present
  + There'll be a coverage report comment
  + PR will be blocked
  
  
  image
  + Here we deliberately ignore check coverage threshold limit (This is my intended behaviour, hence the _"Silently ignoring coverage threshold limit..."_ on line 134)
  + https://github.com/CyberCitizen01/git-proxy/pull/3#issuecomment-1787229271 | [Job run](https://github.com/CyberCitizen01/git-proxy/actions/runs/6707263191/job/18225393944)

- When modified files doesn't have corresponding tests present
  + There'll be no coverage report comment
  + PR will not be blocked (as `nyc` will not even take these files into consideration)  


  image
  + (no comment) | [Job run](https://github.com/CyberCitizen01/git-proxy/actions/runs/6707243459/job/18225331625)

**Comment by JamieSlome** [comment]
@CyberCitizen01 - thank you for the follow-up! 🎉 

Happy to merge if you are @coopernetes...

File: .github/workflows/nodejs.yml
Comment (CyberCitizen01) [inline .github/workflows/nodejs.yml:47]:
```suggestion
      run: npm test

    - name: Test Coverage
      id: test
      run: |
        npm run test-coverage-ci || echo "Silently ignoring coverage threshold limit..."

```

I humbly would like to suggest this change, until we close #240 and revert 222a3dcb523b0b5542d387c4b503f515ac60a9e8. For me, this seemed to be a little outside of the scope, hence thought of asking first.

This will restrict from pushing commits that break existing tests, otherwise everything will be ignored, and it would take manual effort to check whether tests are failing.

**Review by CyberCitizen01 — COMMENTED** [review]
@JamieSlome @coopernetes

**Comment by coopernetes** [comment]
LGTM 👍

File: .github/workflows/nodejs.yml
Comment (JamieSlome) [reply .github/workflows/nodejs.yml:47]: @CyberCitizen01 - I'm a bit confused, why do we need to run the tests twice?

Comment (CyberCitizen01) [reply .github/workflows/nodejs.yml:47]: The following is the current behaviour:
- `npm run test-coverage-ci` (analyse coverage and generate `lcov.info`) -> `npm test` -> `mocha --exit` (run tests and exit with non-zero code if any test fails)
- The expression
  ```bash
  npm run test-coverage-ci || echo "Silently ignoring coverage threshold limit..."
  ```
  basically does, "Run, `npm run test-coverage-ci` and if it fails don't propagate the error code, instead run `echo "Silently ... "` and exit with `0` code

The problem, with the above approach, that fears me is that suppose a PR has modified something that breaks the existing code, it'll go unnoticed even though `npm test` exits with non-zero code, because it is essentially being ignored because of the `||`.

For now, the best suggestion I have is to add an additional step to specifically run tests (`npm test`). We'll only need to do this until the aforementioned conditions are met.

Comment (CyberCitizen01) [reply .github/workflows/nodejs.yml:47]: Running the tests twice is just a quick and temporary fix, until we reach 80% coverage (#240). After that, we'll just revert the suggestions made in https://github.com/finos/git-proxy/pull/213#discussion_r1367778916 thread.

And, too explain further:
Tests can fail, irrespective of the code coverage report generation, it's just that both failing tests and code coverage error out, and since we have removed explicit failure (as suggested earlier), we ignore both aspects of failing tests and code coverage. Running the tests again will make sure that, tests are passing and nothing breaks.

There's one more way to do it without having to run the tests twice - we can try to parse the output of `npm run test-coverage-ci` and try to make sense of it. But this will get messy, and to be honest may not be reliable.

**Comment by CyberCitizen01** [comment]
Hello @JamieSlome 

I wanted to ask and clarify, whether anything from my side is being expected. As it will be lovely if the PR get's merged soon. However I understand if maybe this PR is being regarded as a low priority for now. In that case please let me know if I can do anything to help around here.

Thank you and happy holidays 🎉

**Comment by coopernetes** [comment]
@CyberCitizen01 can you refresh the package-lock.json file by running `npm install` on your branch and committing the changes? I think that should be the last thing remaining and should fix the outstanding CI failure on the [CVE scanning job](https://github.com/finos/git-proxy/actions/runs/7395596341/job/20119059571?pr=213). Then we should be able to merge.

**Comment by JamieSlome** [comment]
@coopernetes @CyberCitizen01 - I've sorted out our failing CI as it was affecting multiple PRs. Are you able to just resolve the outstanding conflict on the `package-lock.json`?

Once that's sorted we'll merge this PR 👍

**Comment by JamieSlome** [comment]
@CyberCitizen01 - just pinging on the above, are we able to get the merge conflict resolved? Very keen to get your fab contributions merged! 👍

**Comment by CyberCitizen01** [comment]
@JamieSlome - Done 🎉

File: .github/workflows/nodejs.yml
Comment (coopernetes) [inline .github/workflows/nodejs.yml:52]: The original action used is based on a deprecated version of `@actions/core` that runs on Node 12.x & generates a deprecation warning[^1]. I haven't seen much activity on the repo and the issue was reported quite awhile ago[^2]. It may be better to use a different coverage reporting action before we merge this. 

Below is a suggested change using [zgosalvez/github-actions-report-lcov](https://github.com/zgosalvez/github-actions-report-lcov) instead. 
```suggestion
    - name: Report code coverage
      uses: zgosalvez/github-actions-report-lcov@v3
      with:
        coverage-files: ./coverage/lcov.info
        github-token: ${{ secrets.GITHUB_TOKEN }}
        update-comment: true
```

[^1]: https://github.blog/changelog/2023-06-13-github-actions-all-actions-will-run-on-node16-instead-of-node12-by-default/
[^2]: https://github.com/romeovs/lcov-reporter-action/issues/55

**Review by coopernetes — COMMENTED** [review]
@JamieSlome @CyberCitizen01 please have a look at the suggested change. We don't want to drag this out too much longer but I am concerned with the deprecation notice being generated.

**Comment by CyberCitizen01** [comment]
@JamieSlome @coopernetes - I've verified that the run of the updated workflow has the expected outcome.

image


https://github.com/CyberCitizen01/git-proxy/pull/3#issuecomment-1903293865

I also went ahead and resolved our conversation (https://github.com/finos/git-proxy/pull/213#discussion_r1377710422).

Please verify and let me know if I can do anything to help.

Lets merge this PR 🦾!

**Comment by coopernetes** [comment]
Seems like the new LCOV report action fails when coverage data is missing. Ugh. 

Can you add `if: success() || failure()` to that specific step? We at least have text-based coverage in the npm step before hand. We can tackle the PR comment getting posted with the coverage info in a subsequent PR so that we at least start testing with coverage. Taking a cue from [other FINOS repos](https://github.com/finos/pylegend/blob/b6d40ad101b5b47621f1c02ff52f200048a9dc52/.github/workflows/build-ci.yml#L53-L71).

@maoo wondering if we can leverage [codecov](https://app.codecov.io/gh/finos) for tracking test coverage outside of GitHub?

**Comment by CyberCitizen01** [comment]
@coopernetes - adding `if: success() || failure()` to a failing step will not skip it. Instead I've just added a step to tackle the situation of an empty `lcov.info` file.

_Rationale for forced push: I thought, if there are no tests for created/modified files `lcov.info` file is not generated altogether, however actually the file does get generated but it is empty._

**Review by coopernetes — APPROVED** [review]

**Comment by coopernetes** [comment]
LGTM - thank you @CyberCitizen01 for the contribution and sticking with it after many iterations!

**Pull request merged**
